pkg/controllermanager/controller/cloudprofile/kubernetesversion/kubernetes_version.go

```diff
@@ -0,0 +1,164 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetesversion
+
+import (
+	"time"
+
+	"github.com/Masterminds/semver"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	controllermgrconfig "github.com/gardener/gardener/pkg/controllermanager/apis/config"
+	"github.com/gardener/gardener/pkg/controllermanager/controller/cloudprofile/helper"
+)
+
+// ReconcileKubernetesVersions reconciles the Spec.Kubernetes.Versions section of the CloudProfile
+// returns true if the CloudProfile has to be updated, the updated CloudProfile or an error
+func ReconcileKubernetesVersions(logger *logrus.Entry, config *controllermgrconfig.CloudProfileControllerConfiguration, cloudProfile *gardencorev1beta1.CloudProfile) (bool, *gardencorev1beta1.CloudProfile, error) {
+	if config == nil || config.KubernetesVersionManagement == nil || config.KubernetesVersionManagement.Enabled == false {
+		return false, cloudProfile, nil
+	}
+
+	profile := cloudProfile.DeepCopy()
+	supportedVersionsMap := make(map[string]helper.HighestSupportedVersion, len(profile.Spec.Kubernetes.Versions))
+	for cloudProfileIndex, version := range profile.Spec.Kubernetes.Versions {
+		currentSemverVersion, err := semver.NewVersion(version.Version)
+		if err != nil {
+			return false, nil, err
+		}
+
+		hasSupportedMinorVersion, err := gardencorev1beta1helper.HasSupportedKubernetesMinorVersion(*config.KubernetesVersionManagement.SupportedKubernetesVersions, profile, *currentSemverVersion)
+		if err != nil {
+			return false, nil, err
+		}
+
+		newerPatchVersionFound, _, err := gardencorev1beta1helper.DetermineLatestKubernetesPatchVersion(profile, version.Version)
+		if err != nil {
+			return false, nil, err
+		}
+		// better readability
+		isLatestPatchVersion := !newerPatchVersionFound
+
+		// TODO: Remove migration code in future version
+		// assumes there is no preview version yet - otherwise no version will be set to supported
+		if version.Classification == nil && isLatestPatchVersion && hasSupportedMinorVersion {
+			// highest patch version of supported minor version -> supported
+			logger.Infof("[KUBERNETES VERSION MANAGEMENT] setting '%s' classification for kubernetes version '%s' that had no classification", gardencorev1beta1.ClassificationSupported, version.Version)
+			version = helper.SetSupportedClassification(version)
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = version
+		} else if version.Classification == nil {
```

Comment:
This `else if` condition looks strange to me. Why are we in the "deprecate case" here?